src/sagas/background-refresh.js

```diff
@@ -106,80 +132,95 @@ function* refreshListPage () {
     return
   }
 
-  const [vmsCount, poolsCount] = yield all([
+  const [vmsResults, poolsResults] = yield all([
     call(function* () {
-      // refresh VMs and remove any that haven't been refreshed
-      if (vmsPage > 0) {
-        const fetchedVmIds = yield fetchVms(Actions.getVmsByCount({
-          count: vmsPage * AppConfiguration.pageLimit,
-        }))
-
-        const vmsIds = yield getIdsByType('vms')
-        const fetchedDirectlyVmIds =
-          (yield all(
-            vmsIds
-              .filter(vmId => !fetchedVmIds.includes(vmId))
-              .map(vmId => call(fetchSingleVm, Actions.getSingleVm({ vmId, shallowFetch: true })))
-          ))
-            .reduce((vmIds, vm) => { if (vm) vmIds.push(vm.id); return vmIds }, [])
-
-        yield put(Actions.removeMissingVms({ vmIdsToPreserve: [...fetchedVmIds, ...fetchedDirectlyVmIds] }))
+      // fetch the VMs we are expecting to be in the pages we have fetched
+      const { internalVms: expectedVms } = yield fetchVms(Actions.getVmsByCount({
+        count: vmsPage * AppConfiguration.pageLimit,
+      }))
+
+      // if any existing VMs are not in expectedVms, fetch them individually
+      const expectedVmIds = new Set(expectedVms.map(vm => vm.id))
+      const existingVmIds = yield getIdSetByType('vms')
```

Comment:
The number of VMs may change in the meantime (i.e. due to infinite scroller fetching new chunk). We should select existing VMs before the call (together with `vmsPage` and other flags).